orchestration/OrchestrationService/CustomScriptExecution.ps1

```diff
@@ -0,0 +1,257 @@
+            # or bash script
+            return $null;
+        }
+    }
+
+    hidden [array] GetArgumentsForBashScript() {
+
+        # Variable to hold the list of arguments to be
+        # passed to the bash script execution
+        $orderedArguments = @();
+
+        # Add the arguments to the array as-is as 
+        # there is no way to verify the order in bash.
+        # We are only converting the hashtable to an 
+        # array
+        $this.arguments.Keys | ForEach-Object {
+            $argumentName = $_;
+            $orderedArguments += $this.arguments[$argumentName];
+        }
+
+        # Return the arguments list
+        return $orderedArguments;
+    }
+
+    hidden [array] GetArgumentsForPowerShellScript() {
+
+        # List of system parameters we can pass to a 
+        # PowerShell script by default
+        $systemParameters = `
+            @(
+                'Verbose',
+                'Debug',
+                'ErrorAction',
+                'WarningAction',
+                'InformationAction',
+                'ErrorVariable',
+                'WarningVariable',
+                'InformationVariable',
+                'OutVariable',
+                'OutBuffer',
+                'PipelineVariable'
+            );
+
+        # Variable to hold the ordered list of arguments to 
+        # be passed to the script execution
+        $orderedArguments = @();
+
+        # Iterate through the list of Parameters accepted by
+        # a script to rearrange the in argument in the right order.
+        (Get-Command $this.command).Parameters.Keys | ForEach-Object {
+            $parameterName = $_;
+
+            # Add the argument to a new array in the right order if
+            # it is passed from the orchestation. Otherwise, add a 
+            # null in its place
+            if($this.arguments.ContainsKey($parameterName) `
+                -and $parameterName -notin $systemParameters) {
+                $orderedArguments += $this.arguments[$parameterName];
+            }
+            elseif($parameterName -notin $systemParameters) {
+                $orderedArguments += $null;
+            }
+        }
+
+        # Return the ordered arguments list
+        return $orderedArguments;
+    }
+
+    hidden [string] RunPowerShellScript([string] $type) {
+
+        # Branch based on the type of PowerShell to
+        # run (i.e script or set of commands)
+        if($type -eq "script") {
+
+            # Get arguments to execute the PowerShell script
+            $argumentsList = `
+                $this.AddArgumentsForExecution($this.scriptType);
+
+            # Pass the script file path and argumentsList to
+            # execute the script
+            return `
+                $this.RunJob($null, $this.command, $argumentsList);
+        }
+        else {
+
+            # Pass the PowerShell commands to execute
+            return `
+                $this.RunJob($this.command, $null, $null);
```

Comment:
Isn't a command capable of receiving arguments as well? for example: ConvertTo-Json is a command that accepts arguments, why are you branching these two?

Reply:
As per our discussion, made the branching for the ScriptBlock vs FilePath parameters when calling Start-Job Cmdlet.